src/network/networkd-dhcp-server.c

```diff
@@ -228,57 +318,24 @@ int dhcp4_server_configure(Link *link) {
-        }
-
-        if (link->network->dhcp_server_emit_ntp) {
-                if (link->network->n_dhcp_server_ntp > 0)
-                        r = sd_dhcp_server_set_ntp(link->dhcp_server, link->network->dhcp_server_ntp, link->network->n_dhcp_server_ntp);
-                else {
-                        if (!acquired_uplink)
-                                uplink = manager_find_uplink(link->manager, link);
-
-                        if (!uplink) {
-                                log_link_debug(link, "Not emitting NTP server information on link, couldn't find suitable uplink.");
-                                r = 0;
-                        } else
-                                r = link_push_uplink_ntp_to_dhcp_server(uplink, link->dhcp_server);
+        if (!dhcp4_server_emit_servers(link) && link->network->dhcp_server_retry_uplink) {
+                usec_t time_now;
 
-                }
+                r = sd_event_default(&link->network->uplink_event);
                 if (r < 0)
-                        log_link_warning_errno(link, r, "Failed to set NTP server for DHCP server, ignoring: %m");
-        }
-
-        if (link->network->dhcp_server_emit_sip) {
-                if (link->network->n_dhcp_server_sip > 0)
-                        r = sd_dhcp_server_set_sip(link->dhcp_server, link->network->dhcp_server_sip, link->network->n_dhcp_server_sip);
-                else {
-                        if (!acquired_uplink)
-                                uplink = manager_find_uplink(link->manager, link);
+                        return log_link_error_errno(link, r, "Failed to allocate event loop: %m");
 
-                        if (!uplink) {
-                                log_link_debug(link, "Not emitting sip server information on link, couldn't find suitable uplink.");
-                                r = 0;
-                        } else
-                                r = link_push_uplink_sip_to_dhcp_server(uplink, link->dhcp_server);
+                r = sd_event_now(link->network->uplink_event, clock_boottime_or_monotonic(), &time_now);
+                if (r < 0)
+                        return r;
 
-                }
+                r = event_reset_time(link->network->uplink_event, &link->network->uplink_timeout_event_source,
+                                     clock_boottime_or_monotonic(),
+                                     time_now + USEC_PER_SEC / 2, 1 * USEC_PER_SEC,
+                                     dhcp4_server_uplink_retry_timeout, link,
+                                     link->network->uplink_event_priority, "dhcp-server-uplink-retry-timeout", true);
                 if (r < 0)
-                        log_link_warning_errno(link, r, "Failed to set SIP server for DHCP server, ignoring: %m");
+                        log_link_debug_errno(link, r, "Failed to set uplink retry timer for DHCP server. Ignoring: %m");
```

Comment:
hmm, can you explain the logic you implemented here? you just ask for a refresh every now and then? why not actually implement this so that whenever the uplink we learn actually changed we propagate stuff, instead of triggering this by time? What am I missing?

Reply:
Sorry for late rely not sure I get this.  How we implement this .  Now we are learning this via a netlink dump.